### ISSUE TITLE
Execute Line in REPL missing

### ISSUE THREAD

**Issue by PatrickHaecker** [julia-vscode/julia-vscode#3654]
Julia code is typically structured in functions. If I haven't missed anything, we have the following code execution possibilities in the REPL:

- active File in REPL
- Code Cell in REPL
- Code Cell in REPL and Move
- Code in REPL
- Code in REPL and Move
- File in REPL

What seems to be completely missing is (in my experience) the most common use case of executing a single line (strictly speaking an expression as it can contain multiple lines). This should also be available together with move.

As a workaround you can highlight a line and run "Execute Code in REPL", but this is rather cumbersome to do for each line of the relevant context.

I also wonder, why the default of "Execute Code in REPL" without any code selection is to execute the surrounding block. Most of the time this is just defining the function and therefore not really executing any "real" code at all and pretty redundant anyway due to Revise which makes the function already available if you save the file.
Wouldn't it make more sense to have the "Execute Current line/expression in REPL" functionality being the default for "Execute Code in REPL? "Execute Surrounding Block in REPL" could then be the new name of the current "Execute Code in REPL" functionality.

**Comment by pfitzseb** [comment]
The difficulty here is finding a good heuristic for determining which exact expression to evaluate. Currently we're running the the biggest non-module expression we can find, which, as you said, evaluates function definitions. 
IMHO that is a very sensible default, because stepping through the lines inside of a function doesn't really make sense without a bunch of setup required (to make arguments available) and is better handled by an actual debugger.

**Comment by pfitzseb** [comment]
You can execute arbitrary code in the local scope via the Debug Console (which admittedly is not quite as nice as the REPL, but it exists).

> If I may suggest a different heuristic 

So you'd want something like this?
```julia
function foo(x,y)
  log(
    x + 1, # cursor here would execute log
    x + y| # only execute this line
  )
end
```

**Comment by PatrickHaecker** [comment]
> You can execute arbitrary code in the local scope via the Debug Console (which admittedly is not quite as nice as the REPL, but it exists).

Yes, you are correct, it is already possible, but it would be so much better if integrated into the REPL.

> ```julia
> function foo(x,y)
>   log(
>     x + 1, # cursor here would execute log
>     x + y| # only execute this line
>   )
> end
> ```
That is a great example. I think I only now completely see your point :-):
> The difficulty here is finding a good heuristic for determining which exact expression to evaluate.

I think your example will work, but it is admittedly a bit non-intuitive and I agree, that `x + 1,` strictly speaking is an incomplete expression (according to `JuliaSyntax.parsestmt` or the REPL), but you could also argue that according to  `:(log(x + 1, x + y)) |> dump` all relevant information of that line is included in a single Expr.

I guess a more intuitive solution would treat `x + 1,` and `x + y` the same. What about the heuristic
- take the longest part of the current line that is a valid expression without narrowing down on any incomplete parentheses (or just special case a leading or trailing comma?)
- extend if necessary to previous or next lines if necessary to get a valid expression and handle multiple expressions as described above

Do you see any non-intuitive behavior there, too?

And by the way: Thanks for all the work you put into the Julia plugin!

**Comment by pfitzseb** [comment]
The simplest implementation of your idea would be "eval the expr that covers all of the line, disregarding trivia". That would however only eval `log` on the first line (because the opening parenthesis gets ignored). 
You can play around with selecting subexpressions with <kbd>Shift-Alt-Right</kbd> and <kbd>Shift-Alt-Left</kbd>, btw.

So yeah, I think there is an implementation of this that is useful, but I'm not entirely sure what it looks like. Special casing commas might be good enough, as you said. 
If you want to try your hand at an implementation of that heuristic (based on JuliaSyntax, probably) then I'd be happy to take a look at that and implement it in the extension if it's useful.
